### PR TITLE
Check for invalid scope during callback

### PR DESCRIPTION
**Pull request by EiNSTeiN-**
Adds an option to validate the granted scopes from shopify.

@dylanahsmith @clayton-shopify for review please

### CODE REVIEW

File: test/integration_test.rb [hunk 1]
```diff
@@ -160,9 +172,9 @@ def sign_params(params)
     params
   end
 
-  def expect_access_token_request(access_token)
+  def expect_access_token_request(access_token, scope)
     FakeWeb.register_uri(:post, "https://snowdevil.myshopify.com/admin/oauth/access_token",
-                         body: JSON.dump(access_token: access_token),
+                         body: JSON.dump(access_token: access_token, scope: scope),
```
Comment (dylanahsmith) [inline test/integration_test.rb]: We are now returning this in the response?  If so, then we should update the documentation (https://docs.shopify.com/api/authentication/oauth) which currently says:

> The server will respond with an access token.
>
> ```
{
  "access_token": "f85632530bf277ec9ac6f649fc327f17"
}
```

Comment (EiNSTeiN-) [inline test/integration_test.rb]: It will once I merge the pending PR in Shopify (you're pinged on it).

I will take care of the docs.

**Comment by EiNSTeiN-** [comment]
I made a small change here so that Shopify can return scopes out of order and this won't break.

**Comment by dylanahsmith** [comment]
:+1:

**Comment by clayton-shopify** [comment]
LGTM

**Comment by EiNSTeiN-** [comment]
Sorry, another update. I just tested this code in a real app and had to make some changes. `fail!` has to be called before calling `super`, because calling it afterwards does not work like I expected. As a result, I added a memoization to `build_access_token` so that it does not make multiple requests to Shopify to fetch the access token.

**Pull request merged**